### DISCUSSION TITLE
How do I use functions without having to sign in?

### DISCUSSION THREAD

**Discussion by squeaktoy** [netdata/netdata#17594] [Q&A]
I'm running Netdata in my own local network, yet the web UI says the following on some tabs like Top:

> **Sign in to Netdata to use this function**
Functions expose sensitive information about your systems and applications. To protect your privacy, Netdata exposes this information only to logged-in users and claimed agents. When viewing Functions directly on a Netdata Agent UI, this information is sent directly from the Netdata Agent to your web browser, without exposing it to any third parties.

I'm on my local network, so why is Netdata stopping me from viewing info about my own computer? How can I disable this requirement to sign in? I don't want to have an account on a third-party server just to view my own computer's information.

It also says something about "claimed agents". I tried the `netdata-claim.sh` script, but it says:
```
This agent was built with --disable-cloud and cannot be claimed
```
So how am I going to use Netdata functions now without Cloud support? I would rather not connect to third-party servers. Anyone have an idea what to do here?

**Comment by ilyam8** [comment]
Hi, @squeaktoy. Not all functions require Cloud connection, see [Available Functions](https://learn.netdata.cloud/docs/top-monitoring-netdata-functions#what-functions-are-currently-available) (`Require Cloud` column).

> It also says something about "claimed agents". I tried the netdata-claim.sh script, but it says:

What is your Netdata version? Can you show `/usr/sbin/netdata -W buildinfo` (or `/opt/netdata/usr/sbin/netdata -W buildinfo` depends on your install prefix).

There was a bug in a nightly version but it was [fixed over a week ago](https://github.com/netdata/netdata/pull/17528).

**Reply by squeaktoy** [reply]
`/usr/sbin/netdata -W buildinfo`:
```
Packaging:
    Netdata Version ____________________________________________ : v1.45.3
    Installation Type __________________________________________ : custom
    Package Architecture _______________________________________ : unknown
    Package Distro _____________________________________________ : unknown
    Configure Options __________________________________________ : dummy-configure-command
Default Directories:
    User Configurations ________________________________________ : /etc/netdata
    Stock Configurations _______________________________________ : /usr/lib/netdata/conf.d
    Ephemeral Databases (metrics data, metadata) _______________ : /var/cache/netdata
    Permanent Databases ________________________________________ : /var/lib/netdata
    Plugins ____________________________________________________ : /usr/libexec/netdata/plugins.d
    Static Web Files ___________________________________________ : /usr/share/netdata/web
    Log Files __________________________________________________ : /var/log/netdata
    Lock Files _________________________________________________ : /var/lib/netdata/lock
    Home _______________________________________________________ : /var/lib/netdata
Operating System:
    Kernel _____________________________________________________ : Linux
    Kernel Version _____________________________________________ : 6.6.28-gentoo-dist
    Operating System ___________________________________________ : Gentoo
    Operating System ID ________________________________________ : gentoo
    Operating System ID Like ___________________________________ : unknown
    Operating System Version ___________________________________ : 2.15
    Operating System Version ID ________________________________ : none
    Detection __________________________________________________ : Mixed
Hardware:
    CPU Cores __________________________________________________ : 8
    CPU Frequency ______________________________________________ : 4000000000
    RAM Bytes __________________________________________________ : 25133178880
    Disk Capacity ______________________________________________ : 6121230925824
    CPU Architecture ___________________________________________ : x86_64
    Virtualization Technology __________________________________ : none
    Virtualization Detection ___________________________________ : none
Container:
    Container __________________________________________________ : unknown
    Container Detection ________________________________________ : none
    Container Orchestrator _____________________________________ : none
    Container Operating System _________________________________ : none
    Container Operating System ID ______________________________ : none
    Container Operating System ID Like _________________________ : none
    Container Operating System Version _________________________ : none
    Container Operating System Version ID ______________________ : none
    Container Operating System Detection _______________________ : none
Features:
    Built For __________________________________________________ : Linux
    Netdata Cloud ______________________________________________ : NO (unavailable)
    Health (trigger alerts and send notifications) _____________ : YES
    Streaming (stream metrics to parent Netdata servers) _______ : YES
    Back-filling (of higher database tiers) ____________________ : YES
    Replication (fill the gaps of parent Netdata servers) ______ : YES
    Streaming and Replication Compression ______________________ : YES (zstd lz4 gzip brotli)
    Contexts (index all active and archived metrics) ___________ : YES
    Tiering (multiple dbs with different metrics resolution) ___ : YES (5)
    Machine Learning ___________________________________________ : YES
Database Engines:
    dbengine ___________________________________________________ : YES
    alloc ______________________________________________________ : YES
    ram ________________________________________________________ : YES
    none _______________________________________________________ : YES
Connectivity Capabilities:
    ACLK (Agent-Cloud Link: MQTT over WebSockets over TLS) _____ : NO
    static (Netdata internal web server) _______________________ : YES
    h2o (web server) ___________________________________________ : YES
    WebRTC (experimental) ______________________________________ : NO
    Native HTTPS (TLS Support) _________________________________ : YES
    TLS Host Verification ______________________________________ : YES
Libraries:
    LZ4 (extremely fast lossless compression algorithm) ________ : YES
    ZSTD (fast, lossless compression algorithm) ________________ : YES
    zlib (lossless data-compression library) ___________________ : YES
    Brotli (generic-purpose lossless compression algorithm) ____ : YES
    protobuf (platform-neutral data serialization protocol) ____ : NO
    OpenSSL (cryptography) _____________________________________ : YES
    libdatachannel (stand-alone WebRTC data channels) __________ : NO
    JSON-C (lightweight JSON manipulation) _____________________ : YES
    libcap (Linux capabilities system operations) ______________ : YES
    libcrypto (cryptographic functions) ________________________ : YES
    libyaml (library for parsing and emitting YAML) ____________ : YES
Plugins:
    apps (monitor processes) ___________________________________ : YES
    cgroups (monitor containers and VMs) _______________________ : YES
    cgroup-network (associate interfaces to CGROUPS) ___________ : YES
    proc (monitor Linux systems) _______________________________ : YES
    tc (monitor Linux network QoS) _____________________________ : YES
    diskspace (monitor Linux mount points) _____________________ : YES
    freebsd (monitor FreeBSD systems) __________________________ : NO
    macos (monitor MacOS systems) ______________________________ : NO
    statsd (collect custom application metrics) ________________ : YES
    timex (check system clock synchronization) _________________ : YES
    idlejitter (check system latency and jitter) _______________ : YES
    bash (support shell data collection jobs - charts.d) _______ : YES
    debugfs (kernel debugging metrics) _________________________ : YES
    cups (monitor printers and print jobs) _____________________ : NO
    ebpf (monitor system calls) ________________________________ : NO
    freeipmi (monitor enterprise server H/W) ___________________ : NO
    nfacct (gather netfilter accounting) _______________________ : NO
    perf (collect kernel performance events) ___________________ : YES
    slabinfo (monitor kernel object caching) ___________________ : YES
    Xen ________________________________________________________ : NO
    Xen VBD Error Tracking _____________________________________ : NO
    Logs Management ____________________________________________ : YES
Exporters:
    AWS Kinesis ________________________________________________ : NO
    GCP PubSub _________________________________________________ : NO
    MongoDB ____________________________________________________ : NO
    Prometheus (OpenMetrics) Exporter __________________________ : YES
    Prometheus Remote Write ____________________________________ : NO
    Graphite ___________________________________________________ : YES
    Graphite HTTP / HTTPS ______________________________________ : YES
    JSON _______________________________________________________ : YES
    JSON HTTP / HTTPS __________________________________________ : YES
    OpenTSDB ___________________________________________________ : YES
    OpenTSDB HTTP / HTTPS ______________________________________ : YES
    All Metrics API ____________________________________________ : YES
    Shell (use metrics in shell scripts) _______________________ : YES
Debug/Developer Features:
    Trace All Netdata Allocations (with charts) ________________ : NO
    Developer Mode (more runtime checks, slower) _______________ : NO
```

**Reply by Myzel394** [reply]
This didn't answer the questions at all.

> How can I disable this requirement to sign in?

How do we do that?

**Reply by squeaktoy** [reply]
After updating to 1..45.4, I still cannot use the `netdata-claim.sh` script

**Reply by gurka** [reply]
> Not all functions require Cloud connection, see [Available Functions](https://learn.netdata.cloud/docs/top-monitoring-netdata-functions#what-functions-are-currently-available) (Require Cloud column).

There it says:
> Since these functions are able to execute routines on the node and due to the potential use cases that they can cover, our concern is to ensure no sensitive information or disruptive actions are exposed through the Agent's API.

I don't understand this. For example, both the Processes and Systemd-journal functions should be strictly read-only - so why is there a concern about "executing routines on the node"? Why doesn't other functions, like Block-devices or Mount-points, have this "problem"?

It feels like some functions are just arbitrary set to require Cloud, just to force people to create and connect their local nodes to a Cloud account...

Edit: additionally, "Processes", which requires Cloud, is handled by the plugin "apps". The documentation of "apps" even says:

> Security
>
> apps.plugin performs a hard-coded function of building the process tree in memory, iterating forever, collecting metrics for each running process and sending them to Netdata. This is a one-way communication, from apps.plugin to Netdata.
> 
> So, since apps.plugin cannot be instructed by Netdata for the actions it performs, we think it is pretty safe to allow it to have these increased privileges.

(https://github.com/netdata/netdata/tree/master/src/collectors/apps.plugin#security)

**Reply by Ou7law007** [reply]
Just like most open source services, if it was actually "for security purposes", they can literally make it possible to register a local account and solve the entire issue once and for all.

**Reply by Ou7law007** [reply]
Having to register your account ON THE CLOUD on its own is a big **f*ck you** to security.

**Reply by MNLierman** [reply]
> Having to register your account ON THE CLOUD on its own is a big **f*ck you** to security.

^^ This. Might seem blunt but how true it is, honestly. I emailed Netdata using their Contact Us option – a few reps on Reddit said invite people wanting to discuss plan options and figure out s solution that works for them to use the contact form. I did so, and gave my argument, including my willingness to agree to create a Netdata Cloud account and link up my devices, IF I could secure my account with additional layers or security and also be able to link up all of my local devices under their free plan. 2 weeks later, I have not received a reply. 

They're milking home users who probably aren't even their big money makers, and all of this new required plan stuff wasn't a thing a few years ago. If I was making money off their products, sure, I'll pay for it then, but I don't. I barely make enough money to pay myself, and some months, I don't even have money to do that. 

One of the reps, somewhere on Reddit, had stated sometime ago, that Netdata is open source and anyone can modify the code, and for that matter, I could create my own account service. Yeah, ok, because I have that level of programming experience to roll my own accounts solution.

**Comment by ribtoks** [comment]
Can we have any way (runtime is preferred, but compile time is also OK) to force-allow access to local functions like systemd-journald? This function can be disabled by default, but overridable with sudo rights.

**Comment by fuad00** [comment]
any updates?

**Comment by Otmanraad** [comment]
for real like forcing us to use their cloud services, like i can just go use a different metric monitoring, well time to uninstall.

**Comment by bits-craft** [comment]
Bump!

**Comment by lyc8503** [comment]
I've also been troubled by this issue and found a possible workaround.

You need to fork the netdata code, modify the `src/database/rrdfunctions-inflight.c` file, remove the relevant permission checks, and then compile it. Like this:

```diff
diff --git a/src/database/rrdfunctions-inflight.c b/src/database/rrdfunctions-inflight.c
index a14234bff431c5..885960b283d26c 100644
--- a/src/database/rrdfunctions-inflight.c
+++ b/src/database/rrdfunctions-inflight.c
@@ -438,48 +438,6 @@ int rrd_function_run(RRDHOST *host, BUFFER *result_wb, int timeout_s,
         return code;
     }
 
-    if(!http_access_user_has_enough_access_level_for_endpoint(user_access, rdcf->access)) {
-
-        if((rdcf->access & HTTP_ACCESS_SIGNED_ID) && !(user_access & HTTP_ACCESS_SIGNED_ID))
-            code = rrd_call_function_error(result_wb,
-                                           "You need to be authenticated via Netdata Cloud Single-Sign-On (SSO) "
-                                           "to access this feature. Sign-in on this dashboard, "
-                                           "or access your Netdata via https://app.netdata.cloud.",
-                                           HTTP_ACCESS_PERMISSION_DENIED_HTTP_CODE(user_access));
-
-        else if((rdcf->access & HTTP_ACCESS_SAME_SPACE) && !(user_access & HTTP_ACCESS_SAME_SPACE))
-            code = rrd_call_function_error(result_wb,
-                                           "You need to login to the Netdata Cloud space this agent is claimed to, "
-                                           "to access this feature.",
-                                           HTTP_ACCESS_PERMISSION_DENIED_HTTP_CODE(user_access));
-
-        else if((rdcf->access & HTTP_ACCESS_COMMERCIAL_SPACE) && !(user_access & HTTP_ACCESS_COMMERCIAL_SPACE))
-            code = rrd_call_function_error(result_wb,
-                                           "This feature is only available for commercial users and supporters "
-                                           "of Netdata. To use it, please upgrade your space. "
-                                           "Thank you for supporting Netdata.",
-                                           HTTP_ACCESS_PERMISSION_DENIED_HTTP_CODE(user_access));
-
-        else {
-            HTTP_ACCESS missing_access = (~user_access) & rdcf->access;
-            char perms_str[1024];
-            http_access2txt(perms_str, sizeof(perms_str), ", ", missing_access);
-
-            char msg[2048];
-            snprintfz(msg, sizeof(msg), "This feature requires additional permissions: %s.", perms_str);
-
-            code = rrd_call_function_error(result_wb, msg,
-                                           HTTP_ACCESS_PERMISSION_DENIED_HTTP_CODE(user_access));
-        }
-
-        dictionary_acquired_item_release(host->functions, host_function_acquired);
-
-        if(result_cb)
-            result_cb(result_wb, code, result_cb_data);
-
-        return code;
-    }
-
     if(timeout_s <= 0)
         timeout_s = rdcf->timeout;
```

Configuring a complete build and packaging environment is cumbersome, but the compilation process can be handled on GitHub Actions. Although the pipeline may fail due to the absence of an official GPG key, you can still download the compiled deb package, like: https://github.com/lyc8503/netdata/actions/runs/17618181095/job/50056891375

Two important points to note:

1. After removing permission checks, anyone can directly access features on the Dashboard. You must add password protection using a reverse proxy or employ methods like SSH forwarding to ensure only you can access the Dashboard.
2. The check for `HTTP_ACCESS_COMMERCIAL_SPACE` is also removed, which appears to bypass the paywall. However, in my brief testing, I found this does not actually do anything. The existing AI insights premium features remain online and still require payment to use. To put it another way, undoubtedly, users have the freedom to modify the Netdata Agent released under the GPL-3.0 license. 
   The only cause for concern is that the Netdata UI (dashboard frontend) is released under a [non-free license](https://app.netdata.cloud/LICENSE.txt), and it mentions `This license allows you to use the Software only to interface with the licensor's other software components, such as Netdata Agents and Netdata Cloud. Any use with replacements for these components is not permitted.`
   The question is, does a Netdata Agent that has only had a few lines of code modified still qualify as a Netdata Agent? Personally, I believe it does. This is what open-source software is all about.

Regardless, I've been using Netdata on my HomeLab for over seven years now. Its diagnostic capabilities and detailed metrics help me quickly resolve complex issues. I really like it. 

Though recent changes, including the tight coupling with Netdata Cloud and the removal of the open-source v1 dashboard, have left me somewhat disappointed. We hope Netdata will become more user-friendly for individual users in terms of its product and design, rather than rushing to promote Netdata Cloud and expensive subscription plans.

**Reply by fuad00** [reply]
nice one! people certainly need a docker image of that!

**Reply by lyc8503** [reply]
> nice one! people certainly need a docker image of that!

I found some metrics missing when running in docker (even after mounting all required files), so I install Netdata directly on host.

If you'd like to have a docker image, you can pull the Netdata repo, make the modification above and build it locally. It should be much easier to build locally with Docker. Dockerfile is located at `packaging/docker/Dockerfile`.

**Comment by budius** [comment]
I'm happy I didn't install locally and just ran it as a docker container for testing.
Now I can just 'docker compose down' and go find a different solution.

Thanks for the reports and confirm it's not really possible.

**Comment by The1hauntedX** [comment]
Good find, @lyc8503! It looks like a simpler change can instead be made to `src/web/api/http_auth.h`
```diff
diff --git a/src/web/api/http_auth.h b/src/web/api/http_auth.h
index 0b01fdb1e..6c6f67802 100644
--- a/src/web/api/http_auth.h
+++ b/src/web/api/http_auth.h
@@ -15,7 +15,7 @@ time_t bearer_create_token(nd_uuid_t *uuid, HTTP_USER_ROLE user_role, HTTP_ACCES
 bool web_client_bearer_token_auth(struct web_client *w, const char *v);
 
 static inline bool http_access_user_has_enough_access_level_for_endpoint(HTTP_ACCESS user, HTTP_ACCESS endpoint) {
-    return ((user & endpoint) == endpoint);
+    return true;
 }
```

This also unlocks the ability to mange collector configs from the UI as well.

**Comment by vhbui02** [comment]
I want to give thanks to both @The1hauntedX and @lyc8503 answers, and create an automated workflow for it:
- First, fork the netdata/netdata repository.
- Second, make changes to `src/web/api/http_auth.h` according to @The1hauntedX 
- Third, create a GitHub PAT with read-write permission on "Contents" and put it in the fork's repo secret via GitHub Web UI.
- Fourth, add a GitHub Actions workflow file to automate fork syncing (make sure to add your name and email at step 1):

```yml
# Cre: https://github.com/orgs/community/discussions/153608#discussioncomment-12453674
name: Sync Fork with Upstream
on:
  schedule:
    - cron: "0 0 * * *" # Runs daily at midnight UTC
  workflow_dispatch: # Allows manual triggering

jobs:
  sync:
    runs-on: ubuntu-latest
    steps:
      - name: Configure Git
        run: |
          git config --global user.email "insert your email here"
          git config --global user.name "insert your name here"

      - name: Checkout Forked Repository
        uses: actions/checkout@v6
        with:
          token: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
          fetch-depth: 0

      - name: Sync Upstream & Apply Patch
        run: |
          # 1. Fetch Upstream
          git remote add upstream https://github.com/netdata/netdata.git
          git fetch upstream master

          # 2. Hard Reset: Wipe local history to match Upstream exactly
          # This removes yesterday's patch commit so we don't get conflicts
          git checkout master
          git reset --hard upstream/master

          # 3. Apply the Hack (sed)
          echo "Applying auth bypass patch..."
          sed -i 's/return ((user & endpoint) == endpoint);/return true;/' src/web/api/http_auth.h

          # 4. Commit the changes
          git add src/web/api/http_auth.h
          git commit -m "feat: bypass Netdata Cloud sign-in which is required to access dashboard features"

          # 5. Force Push
          # We force push because we rewrote history (replaced yesterday's commit with today's)
          git push origin master --force
```
- Fifth, commit the changes.
- Sixth, copy this script into `/usr/local/bin/netdata-custom-updater.sh` and change `$REPO_OWNER`:

> [!NOTE]
> If you're not using Ubuntu 24.04, refer to the artifact list at `https://github.com/lyc8503/netdata/actions/runs/17618181095/` and change `$ARTIFACT_NAME` accordingly:

```
#!/usr/bin/env bash

set -euo pipefail

REPO_OWNER="vhbui02"
REPO_NAME="netdata"
ARTIFACT_NAME="ubuntu-24.04-amd64-packages"

# https://docs.github.com/en/rest/actions/artifacts?apiVersion=2022-11-28#list-artifacts-for-a-repository
ARTIFACT_URL=$(curl -fsL \
    -H "Accept: application/vnd.github+json" \
    -H "Authorization: Bearer ${GITHUB_TOKEN}" \
    -H "X-GitHub-Api-Version: 2022-11-28" \
    "https://api.github.com/repos/${REPO_OWNER}/${REPO_NAME}/actions/artifacts?per_page=100&page=1&name=${ARTIFACT_NAME}" \
    | jq -r '.artifacts.[].archive_download_url')

if [ -z "${ARTIFACT_URL}" ] || [ "${ARTIFACT_URL}" == "null" ]; then
    echo "Error: Artifact not found!"
    exit 1
fi

curl -fsL \
    -H "Accept: application/vnd.github+json" \
    -H "Authorization: Bearer ${GITHUB_TOKEN}" \
    -H "X-GitHub-Api-Version: 2022-11-28" \
    -o /tmp/netdata.zip "${ARTIFACT_URL}"

# -o: overwrite files without prompting
# -qq: quiet mode
unzip -o -qq /tmp/netdata.zip -d /tmp/netdata

# -o Dpkg::Options::="--force-confold" : Keep old config (default=N)
# -o Dpkg::Options::="--force-confdef" : Use default if no modified file exists
if ! (command -v netdata >/dev/null); then
    sudo apt-get install \
        -y \
        -o Dpkg::Options::="--force-confdef" \
        -o Dpkg::Options::="--force-confold" \
        /tmp/netdata/netdata*.deb
else
    sudo apt-get install \
        -y \
        -o Dpkg::Options::="--force-confdef" \
        -o Dpkg::Options::="--force-confold" \
        --only-upgrade \
        /tmp/netdata/netdata*.deb
fi

# Cleanup
rm -f /tmp/netdata.zip 
rm -rf /tmp/netdata
```
- Seventh, mask the built-in systemd updater:

```sh
sudo systemctl mask --now netdata-updater.timer
sudo systemctl mask --now netdata-updater.service
```

- Eighth, create your own systemd updater (remember to update GITHUB_TOKEN inside service file and set its permission to 600, or you can use EnvironmentFile= entry with hardened env file instead):

```
# /etc/systemd/system/netdata-custom-updater.timer
[Unit]
Description=Daily auto-updates for Netdata
RefuseManualStart=no
RefuseManualStop=no

[Timer]
Persistent=false
OnCalendar=daily
Unit=netdata-custom-updater.service

[Install]
WantedBy=timers.target
```

```
# /etc/systemd/system/netdata-custom-updater.service
[Unit]
Description=Daily auto-updates for Netdata
RefuseManualStart=no
RefuseManualStop=yes

[Service]
Type=oneshot
Environment="GITHUB_TOKEN=insert-your-token-here"
ExecStart=/usr/local/bin/netdata-custom-updater.sh
```

- Ninth, enable the auto updater:

```
sudo systemctl enable --now netdata-custom-updater.timer
```

- Tenth, trigger the auto updater manually for initial installation:

```
sudo systemctl start netdata-custom-updater.service
```

P/s: This 10-step workflows is my fight to vendor lock-in and attempt to try to make us stupid by saying cloud is necessary for basic auth.

EDIT: I've updated a better CI/CD job that can avoid conflicts even if `src/web/api/http_auth.h` changed in the future (which hasn't changed in the last 2 years).

**Reply by sashwathn** [reply]
@vhbui02 : Satya here and I head Product at Netdata.

Thanks for the discussion — and here's a response just to clear things up 🙂

What's being talked about here isn't a bug or something that can be "configured away". The functionality in question is part of the **Netdata Cloud / Enterprise UI**, which is covered by the Netdata Cloud UI license:
https://app.netdata.cloud/LICENSE.txt

That license doesn't allow bypassing or disabling licensing or access controls, so we can't support or recommend doing that.

It's also worth stressing that Netdata isn't "free or enterprise-only". We offer several options, depending on the use case:

- **Netdata Agent** is still GPL and can always be used locally without Netdata Cloud
- **Netdata Cloud SaaS** for a fully managed experience
- **Netdata Cloud On-Prem** for self-hosted environments
- **Netdata Cloud HomeLab**, which is intentionally very generous for personal/small setups, and currently costs **$90/year**

If you want the advanced UI features without limitations, one of the Netdata Cloud plans — especially HomeLab for non-commercial use — is the supported way to get them.

We're happy to help clarify what's included in each option or point you to a setup that fits your needs. Discussions or instructions about bypassing licensing, however, are out of scope and won't be supported here.

**Reply by DDoSolitary** [reply]
Thanks, for clarifying that modifying GPL licensed code is "bypassing or disabling licensing or access controls". 🙂

**Reply by sashwathn** [reply]
@DDoSolitary: There is a reason we have created multiple offerings - to try and fit in different requirements - Community, Homelab and Businesses. 
The Netdata Dashboard and any access controls through it falls under a separate license.

I understand your perspective but we do not want to continue the discussion on this thread discussing about bypassing our access controls (hope you understand).

I am more than happy to get on a call and discuss the best way forward. We are in fact ideating an Enterprise Agent offering and it will be good to get some feedback from you. 😄
